From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Rose <jeremya@chromium.org>
Date: Wed, 2 Sep 2020 09:40:18 -0700
Subject: add CredentialsMode::kOmitCookies_Electron

This adds a new value to the CredentialsMode enum in the fetch API. The
new value emulates an older behavior in Electron, where sending a
request would send along authentication credentials in response to a
401, but would not send cookies. That behavior became no longer
expressible with https://chromium-review.googlesource.com/c/chromium/src/+/2367394.

In future, Electron will remove this behavior and only support either
kOmit (i.e. send neither cookies or auth) or kInclude (send both cookies
and auth). At that time, this patch should be removed.

diff --git a/services/network/public/cpp/cors/cors.cc b/services/network/public/cpp/cors/cors.cc
index 7189eaa1871ddd13d998e4c0ccec7c6b062d5237..665b804081f5b9cd39ffe9e5a88895a18b573c33 100644
--- a/services/network/public/cpp/cors/cors.cc
+++ b/services/network/public/cpp/cors/cors.cc
@@ -622,6 +622,7 @@ bool CalculateCredentialsFlag(mojom::CredentialsMode credentials_mode,
     case network::mojom::CredentialsMode::kSameOrigin:
       return response_tainting == network::mojom::FetchResponseType::kBasic;
     case network::mojom::CredentialsMode::kInclude:
+    case network::mojom::CredentialsMode::kOmitCookies_Electron:
       return true;
   }
 }
diff --git a/services/network/public/mojom/fetch_api.mojom b/services/network/public/mojom/fetch_api.mojom
index 8b3f2f16df17f4f1e38615600e9624997e6742c3..892c5b7ba180717eccd15fe1c3e35c2b7ba0e94b 100644
--- a/services/network/public/mojom/fetch_api.mojom
+++ b/services/network/public/mojom/fetch_api.mojom
@@ -69,6 +69,11 @@ enum CredentialsMode {
   // and continue the handshake without sending one if requested.
   kOmit,
 
+  // TODO(jeremy): This emulates an older behavior that Electron used, to allow
+  // omitting cookies while still sending auth credentials. It will eventually
+  // be removed.
+  kOmitCookies_Electron,
+
   kSameOrigin,
   kInclude,
 
diff --git a/third_party/blink/renderer/core/fetch/request.cc b/third_party/blink/renderer/core/fetch/request.cc
index 3f45de671ed4cea7f94ea1a8995dc4489451cb35..263f7eef8ce5e3accc863fe5d088db98eca8b299 100644
--- a/third_party/blink/renderer/core/fetch/request.cc
+++ b/third_party/blink/renderer/core/fetch/request.cc
@@ -864,6 +864,7 @@ String Request::credentials() const {
   // mode:"
   switch (request_->Credentials()) {
     case network::mojom::CredentialsMode::kOmit:
+    case network::mojom::CredentialsMode::kOmitCookies_Electron:
     case network::mojom::CredentialsMode::kOmitBug_775438_Workaround:
       return "omit";
     case network::mojom::CredentialsMode::kSameOrigin:
